refactor(modules): introduce modularization

mikewegele opened the pull request.

mikewegele committed:
chore: run format (#77)

koganesi committed:
Merge branch 'DEV-2/IAV-FF-chore-328' into iav-ff-feature-imprint-344

koganesi committed:
refactor: translation usage in privacyPolicyText component

Jonas-Krahn committed:
chore: Fix file header

Commit:
chore: run format on docs

mikewegele committed:
chore: run format on docs (#78)

Commit:
chore: run format on docs

mikewegele committed:
chore: run format on docs (#79)

Commit:
chore: remove {}

Commit:
chore: remove {}

mikewegele committed:
chore: remove {} (#80)

Commit:
chore: commit to base ref

Commit:
chore: npm run format

Commit:
chore: try npm run format

mikewegele committed:
Merge pull request #5 from mikewegele/DEV-2/IAV-FF-chore-328-mike

chore: try npm run format

Jonas-Krahn committed:
chore: further documentation

Commit:
chore: try npm run format

mikewegele committed:
Merge pull request #6 from mikewegele/DEV-2/IAV-FF-chore-328-mike

chore: try npm run format

Jonas-Krahn committed:
Merge branch 'DEV-2/IAV-FF-chore-328' of https://github.com/iavofficial/IAVFrontendFramework into DEV-2/IAV-FF-chore-328

Commit:
chore: try npm run format

Commit:
chore: add fi

mikewegele committed:
Merge pull request #7 from mikewegele/DEV-2/IAV-FF-chore-328-mike

Dev 2/iav ff chore 328 mike

Commit:
chore: add fi

Commit:
chore: add fi checkout

Commit:
chore: add fi checkout

mikewegele committed:
Merge pull request #8 from mikewegele/DEV-2/IAV-FF-chore-328-mike

Dev 2/iav ff chore 328 mike

Commit:
chore: add fi checkout

Commit:
chore: add fi checkout

mikewegele committed:
Dev 2/iav ff chore 328 format (#81)

mikewegele committed:
chore(format): apply automated formatting changes

Jonas-Krahn committed:
chore: further implementation and fix of router

Jonas-Krahn committed:
Merge branch 'DEV-2/IAV-FF-chore-328' of https://github.com/iavofficial/IAVFrontendFramework into DEV-2/IAV-FF-chore-328

koganesi committed:
feat: update legal documents components in dev_project

koganesi committed:
Merge branch 'DEV-2/IAV-FF-chore-328' of https://github.com/iavofficial/IAVFrontendFramework into iav-ff-feature-imprint-344

koganesi committed:
Merge branch 'DEV-2/IAV-FF-chore-328' of https://github.com/iavofficial/IAVFrontendFramework into iav-ff-feature-imprint-344

Jonas-Krahn committed:
docs: Further documentation and renaming

Jonas-Krahn committed:
chore: Further documentation and refactoring

koganesi committed:
fix: adjust legal documents' header's padding

koganesi committed:
feat(imprint): separate imprint and privacy policy (#73)

koganesi committed:
chore(format): apply automated formatting changes

Jonas-Krahn committed:
chore: Generated new SBOMs

Jonas-Krahn committed:
Merge branch 'DEV-2/IAV-FF-chore-328' of https://github.com/iavofficial/IAVFrontendFramework into DEV-2/IAV-FF-chore-328

Commit:
Merge remote-tracking branch 'mike/DEV-2/IAV-FF-chore-328' into DEV-2/IAV-FF-chore-328

Commit:
Merge remote-tracking branch 'origin/DEV-2/IAV-FF-chore-328' into DEV-2/IAV-FF-chore-328

mikewegele commented on packages/shared/src/modules/default_modules/auth/dummyAuthenticatorModule.tsx:
```diff
@@ -0,0 +1,99 @@
+      reducers: {
+        login: (state, action: PayloadAction<string>) => {
+          state.isLoading = true;
+          state.hasAuthenticated = true;
```
has to be renamed to draft
- draft: indicates that you are working on a temporary copy of the state (mutable).
- state: would imply that you are working directly on the immutable state, which is not the case.

It is therefore helpful to name the variable draft to make it clear that you are working with a "draft" of the state that will later be transferred to the immutable state.

I know it's just nuances, but it's very, very important to understand the idea behind it.

Jonas-Krahn replied:
This is a good idea. However, although state changes are immutable in Redux using Immer, the Redux documentation also uses "state":

https://redux-toolkit.js.org/api/createSlice

I think the reason may be that many Frontend developers don't get the concept of immutability and so they don't want to confuse them.

mikewegele commented on packages/shared/src/modules/module_orchestration/util/mergeModules.ts:
```diff
@@ -0,0 +1,102 @@
+  userNonStoreModules: TUserNonStoreModules;
+}) => {
+  type TMergedFrameworkStoreModules = MergeModules<
+    DefaultStoreModules,
```
It would be better to always deflect the return type. Typing is the be-all and end-all in Typescript. You have done this very well in parts, but there are still a few places where something is missing :)

Before
![image](https://github.com/user-attachments/assets/8277ae66-1f6e-4be0-b4ba-3a9fb1827104)


After
![image](https://github.com/user-attachments/assets/7befe1a6-81e1-403b-a423-9ccfc48b9e50)

Its much easier to debug and understand the code

mikewegele commented on packages/aws-authenticator/src/components/awsAuthenticationView.tsx:
```diff
@@ -18,32 +18,82 @@
+>(
+  props: AuthenticationViewProps,
+) => {
+  const {modules} = useModuleContext<TModules>();
```
This does not work. I can still access to router, userTest, userModules etc...

I think we dont need generic components. Its extremely impractical to use if you want to propagate generic dependet parameters. 

A better way instead of useModuleContext<TModules>();
would be solutions like 
useModuleContext("auth", "router"), useModuleContext("auth", "internationlizer"), useModuleContext("router") etc...
which is much better and does not require generic components. I have not changed it because useModuleContext is used in much complex situations and would cause a deep dive and many refactorings i guess.

Its the same idea behind my useApiState function with deconstructing the parameter.
http://github.com/mikewegele/react/blob/master/src/screens/Project.tsx#L25

Jonas-Krahn replied:
The problem is:
As of creation time of the Framework we don`t known which exact module types the user will use. Because of this we cannot use the correct types for the creation of the context, only the types we know have to exist --> The minimal default implementations.

Because of this we need a generic useModuleContext in order to let it return the correct type.
Furthermore: I already thought of a Hook useModules where you just pass the generic module type and the keys like you suggested. However, with TS it is not possible to ensure that all the keys have to exist inside the generic module type. Try your luck, if you don`t think so. I would be happy to be proven wrong. :D

What do you mean by we don't need generic components? Is this just meant for the AWSAuthenticationView?

mikewegele commented on packages/aws-authenticator/src/components/awsAuthenticationView.tsx:
```diff
@@ -18,32 +18,82 @@
+    ) ?? "";
+
+  const isLoading = useTypedSelector(
+    (state) => state[MandatoryModuleNames.Authenticator].isLoading,
```
const authState = useTypedSelector(state => state.auth);

    const isNewPasswordRequired = authState.extras.isNewPasswordRequired;
    const loginError = authState.extras.loginError ?? "";
    const isLoading = authState.isLoading;
    
    The approach is better because:
- You call `useTypedSelector` once, avoiding repetitive calls.
- Grouping related state values together makes the code clearer.
- Calling `useTypedSelector` once reduces unnecessary re-renders.

Jonas-Krahn replied:
I think this is partially correct: While it may save some lines of code, selecting more values than you may need may result in more rerenders. In Redux change of object is tracked by comparison of references, that`s why we use immutable updates. If you selected the whole ".auth." you will get rerenders whenever something in auth changes, although you may not need it.

mikewegele commented on packages/shared/src/constants/moduleNames.ts:
```diff
@@ -0,0 +1,28 @@
+
+export const MandatoryModuleNames = {
+  Authenticator: "auth",
+  Router: "router",
```
why not 
export type MandatoryModuleNames = "auth" | "router" | "internationalizer"

Jonas-Krahn replied:
Because it should also be used as a value. Is this possible by using a type?

mikewegele commented on packages/shared/src/contexts/moduleContext.ts:
```diff
@@ -0,0 +1,55 @@
+  TModules = AllDefaultModules,
+>(
+  key: K,
+) => {
```
export type MandatoryModuleValues = typeof MandatoryModuleNames[keyof typeof MandatoryModuleNames];

export const useModule = <
    K extends MandatoryModuleValues,
    TModules = AllDefaultModules,
>(
    key: K,
) => {
    const {modules} = useModuleContext<TModules>();
    return modules[key];
};

Only the permitted values ("auth", "router", "internationalizer") are suggested when using useModule. This makes the code more secure, as the developer can only work with these permitted values. The TypeScript editor will also make suggestions for these values.

![image](https://github.com/user-attachments/assets/add7cd06-705f-4d0b-985b-d7ae15e755af)

![image](https://github.com/user-attachments/assets/632e7870-7f8c-419b-9019-d8775b74a762)

Jonas-Krahn replied:
However, you are just extending MandatoryModuleNames here. This is not correct as useModule should also be usable for user modules. Or am i wrong here?

koganesi committed:
fix(imprint): adjust layout and remove '&' separator, add scrollability

Commit:
docs: change title and image

Commit:
chore: remove webkit from navbar.css

Commit:
chore: npm run format

Commit:
fix: height inherit for navbar.tsx

Commit:
fix: height inherit for navbar.tsx

Commit:
chore: npm run format

koganesi committed:
fix(imprint): adjust layout and remove '&' separator, add scrollability (#83)

Commit:
Merge remote-tracking branch 'origin/DEV-2/IAV-FF-chore-328' into DEV-2/IAV-FF-chore-328

Commit:
docs: add version 1.4.0

koganesi committed:
feat: add dynamic legal document configuration

Introduced a new prop to configure legal document links dynamically.
Updated UI to display legal document links conditionally based on visibility settings.
Added support for rendering custom components for legal document pages.
Removed static configurations for imprint and privacy policy.
Updated route handling to dynamically register paths for legal documents.

koganesi committed:
docs: enhance documentation to guide users in configuring legal documents.

koganesi committed:
feat: add example legal docs in dev project

koganesi committed:
Merge branch 'DEV-2/IAV-FF-chore-328' into IAV-FF-feature-custom-legal-docs-367

koganesi committed:
refactor: apply suggested changes from review

Commit:
docs: versionMappings.ts change list

Commit:
docs: working with search function

koganesi committed:
feat: allow legal doc links to wrap when space is limited

Commit:
chore: change import from index

Commit:
chore: change import from index

Commit:
docs: implement searchHeaders.tsx

Commit:
docs: adapt getPages function

Commit:
docs: adapt getPages function

Commit:
docs: adapt searchHeaders.tsx styling

Commit:
docs: add badges for new elements

koganesi committed:
feat: add dynamic legal document configuration (#86)

koganesi committed:
chore(format): apply automated formatting changes

koganesi committed:
feat: add fonts inclusion to SBOM generation

Commit:
chore: npm run format on push

Commit:
Merge remote-tracking branch 'origin/DEV-2/IAV-FF-chore-328' into DEV-2/IAV-FF-chore-328-search

github-actions[bot] committed:
chore(format): apply automated formatting changes

Commit:
Merge remote-tracking branch 'origin/DEV-2/IAV-FF-chore-328-search' into DEV-2/IAV-FF-chore-328-search

Commit:
chore: add command npm run build docs

koganesi committed:
feature: run add_fonts_to_sbom.sh on build and test (will be removed)

koganesi committed:
feature: add workflow_dispatch for build and test workflow

koganesi committed:
revert changes

koganesi committed:
revert changes

mikewegele committed:
feat(docs): add search function (#90)

Commit:
docs: update mistakes in pageInterface.tsx

koganesi committed:
test pipeline execution on push

koganesi committed:
Add SBOM workflow for Google Fonts integration

Commit:
docs: update mistakes in pageInterface.tsx

koganesi committed:
feat(sbom_fonts): update script, update SBOM files encoding

koganesi committed:
feat(sbom_fonts): remove google fonts entry from sharde/sbom to test pipeline

koganesi committed:
feat(sbom_fonts): update pipeline to commit changes after script execution

actions-user committed:
Added Roboto-Regular font entry to SBOM files

koganesi committed:
feat(sbom_fonts): integrate sbom update into release pipeline

koganesi committed:
feat(sbom_fonts): update bash script to add entry into the "relationships" array

koganesi committed:
feat: update CI pipeline for SBOM commits with semantic support

koganesi committed:
Add Google Fonts Entry to SBOM using the Pipeline (225) (#91)

Commit:
Merge remote-tracking branch 'origin/DEV-2/IAV-FF-chore-328' into DEV-2/IAV-FF-chore-328

koganesi committed:
create foss check workflow (testing)

koganesi committed:
integrate npm install into pipeline

koganesi committed:
modify npm audit step

koganesi committed:
mdoify licensee intsallation scope to user

koganesi committed:
correct license compliance check

koganesi committed:
extend the foss check pipeline

koganesi committed:
Introduce testing vulnerability in lodash (4.17.19)

- testing the FOSS Check pipeline

koganesi committed:
Test license compliance using Mapbox SDK with incompatible license

koganesi committed:
Test dependency license enforcement

koganesi committed:
fix: refine the dependency licence checker logic

koganesi committed:
fix: refine the dependency licence checker logic

koganesi committed:
fix: refine the dependency licence checker logic

koganesi committed:
fix: refine the dependency licence checker logic

koganesi committed:
fix: refine the dependency licence checker logic

koganesi committed:
fix: try another approach for the Dependency License Compliance Check

koganesi committed:
fix: try another approach for the Dependency License Compliance Check

koganesi committed:
fix: workflow syntax

koganesi committed:
fix: workflow syntax

koganesi committed:
fix: workflow syntax

Commit:
chore: npm run format

Commit:
chore: remove version banner

Commit:
chore: add readmes for projects

Commit:
chore: adapt docs readme

github-actions[bot] committed:
chore(format): apply automated formatting changes

koganesi committed:
fix: try another approach for licenses audit

koganesi committed:
fix: try another approach for licenses audit

koganesi committed:
fix: try different package.json config

koganesi committed:
fix: try different approach for foss check

koganesi committed:
fix: remove made-up-module

koganesi committed:
fix: actions versions

koganesi committed:
fix: update Vulnerability level cutoff

koganesi committed:
chore(ci): full FOSS compliance (SBOM, CVEs, license policy, SARIF)

koganesi committed:
chore(ci): update actions versions

koganesi committed:
chore(ci): fix errors

koganesi committed:
chore(ci): fix report issues

koganesi committed:
chore(ci): fix gh actions versions

koganesi committed:
chore(ci): fix errors

koganesi committed:
chore(ci): update sbom generation

koganesi committed:
chore(ci): update sbom generation

koganesi committed:
chore(ci):add .ort.yml

koganesi committed:
chore(ci): update the ORT License & Policy Check step

koganesi committed:
chore(ci): update the ORT License & Policy Check step

koganesi committed:
chore(ci): update the ORT License & Policy Check step

koganesi committed:
chore(ci): update the ORT License & Policy Check step

koganesi committed:
chore(ci): update the ORT License & Policy Check step

koganesi committed:
chore(ci): isolate ort-check into a separate job

koganesi committed:
chore(ci): update the ort-check job

koganesi committed:
chore(ci): update ort.yml to ignore local workspace packages

koganesi committed:
chore(ci): update the ort-check job

koganesi committed:
chore(ci): fix typo in foss_check.yml

koganesi committed:
chore(ci): update the ort-check job

koganesi committed:
chore(ci): fix typo in foss_check.yml

koganesi committed:
chore(ci): update the ort-check job

koganesi committed:
chore(ci): update the ort-check job

koganesi committed:
chore(ci): update the ort-check job add npm install step

koganesi committed:
chore(ci): update the ort-check job add ignore_dev_dependencies

koganesi committed:
chore(ci): update the ort-check job

koganesi committed:
chore(ci): fix ort.yml

koganesi committed:
reset last changes

koganesi committed:
chore(ci): update the ort-check job

koganesi committed:
chore(ci): update the ort-check job

koganesi committed:
chore(ci): update ort.yml

koganesi committed:
chore(ci): update ort global config

koganesi committed:
chore(ci): update ignored paths

koganesi committed:
chore(ci): simplify run ort step

koganesi committed:
chore(ci): test different glob version

koganesi committed:
chore(ci): reset new linking for local packages

koganesi committed:
chore(ci): reset new linking for local packages

koganesi committed:
chore(ci): try cleaning cache before installing deps

koganesi committed:
chore(ci): add repository config for ORT to ignore certain paths

koganesi committed:
chore(ci): cleanup config files

koganesi committed:
chore(ci): remove unnecessary steps

koganesi committed:
chore(ci): add licence rules into foss check pipeline

koganesi committed:
chore(ci): Add shields package (GPL license) for FOSS Check test

koganesi committed:
chore(ci): Add shields package (GPL license) for FOSS Check test

koganesi committed:
chore(ci): Add more gpl packages for FOSS Check test

koganesi committed:
chore(ci): Add more gpl packages for FOSS Check test

koganesi committed:
chore(ci): add a step to upload generated sbom

koganesi committed:
chore(ci): add a job to generate sbom for each package

koganesi committed:
chore(ci): fix the script iterating through packages

koganesi committed:
chore(ci): fix the script iterating through packages

koganesi committed:
chore(ci): fix the script iterating through packages

koganesi committed:
chore(ci): fix the script iterating through packages

koganesi committed:
chore(ci): fix typo

koganesi committed:
chore(ci): fix the script iterating through packages

koganesi committed:
chore(ci): fix the script iterating through packages

koganesi committed:
chore(ci): fix the script iterating through packages

koganesi committed:
chore(ci): fix the script iterating through packages

koganesi committed:
chore(ci): debug

koganesi committed:
chore(ci): debug

koganesi committed:
chore(ci): debug

koganesi committed:
chore(ci): debug

koganesi committed:
chore(ci): debug

koganesi committed:
chore(ci): debug

koganesi committed:
chore(ci): debug

koganesi committed:
chore(ci): try new approach for per-package-sbom job

koganesi committed:
chore(ci): add step to install Anchore tools

koganesi committed:
chore(ci): debug

koganesi committed:
chore(ci): debug

koganesi committed:
chore(ci): debug

koganesi committed:
chore(ci): upload grype sarif scan results to github's code scanning alerts

koganesi committed:
chore(ci): refine pipeline

koganesi committed:
chore(ci): fix syntax error

koganesi committed:
chore(ci): fix syntax error

koganesi committed:
chore(ci): check if sbom has changed

koganesi committed:
chore(ci): temporarily skip the ort-check job

koganesi committed:
chore(ci): further debug

koganesi committed:
chore(ci): further debug

koganesi committed:
chore(ci): further debug

koganesi committed:
chore(ci): further debug

koganesi committed:
chore(ci): further debug

koganesi committed:
chore(ci): final version of the working pipeline

koganesi committed:
chore(ci): set GitHub identity

actions-user committed:
chore(ci): add SBOMs for workspaces

koganesi committed:
chore(ci): remove GPL packages

actions-user committed:
chore(ci): add SBOMs for workspaces

koganesi committed:
chore(ci): remove GPL packages from package-lock

koganesi committed:
chore(ci): debug SBOM generation

koganesi committed:
chore(ci): add debug logs

koganesi committed:
chore(ci): try to fix the paths issue

actions-user committed:
chore(ci): add SBOMs for workspaces

koganesi committed:
chore(ci): try to fix the paths issue

actions-user committed:
chore(ci): add SBOMs for workspaces

koganesi committed:
chore(ci): try to fix the paths issue

koganesi committed:
Merge branch 'IAV-FF-feature-foss-check-integration-192' of https://github.com/iavofficial/IAVFrontendFramework into IAV-FF-feature-foss-check-integration-192

actions-user committed:
chore(ci): add SBOMs for workspaces

koganesi committed:
chore(ci): fix the paths issue

koganesi committed:
Merge branch 'IAV-FF-feature-foss-check-integration-192' of https://github.com/iavofficial/IAVFrontendFramework into IAV-FF-feature-foss-check-integration-192

koganesi committed:
chore(ci): remove wrongfully generated files

koganesi committed:
chore(ci): include current branch into foss check workflow

koganesi committed:
fix(vulnerabilities): bump vite from 5.3.1 to 6.3.5

koganesi committed:
fix(vulnerabilities): rerun workflow

koganesi committed:
fix(vulnerabilities): bump react-syntax-highlighter from 15.6.1 to ^15.6.1 for docs

koganesi committed:
fix(vulnerabilities): override prismjs version from 1.27.0 to 1.30.0 for docs

mikewegele committed:
chore(readme): add readmes for sub projects (#104)

koganesi committed:
fix(vulnerabilities): fix critical / high / medium vulnerabilities found by Grype SARIF scan introduced in the new FOSS Check workflow (383) (#111)

koganesi committed:
fix(ci): try to fix the build issue

koganesi committed:
feat(build): adapt build process to include shared assets in build-lifecycle

koganesi committed:
chore(ci): uncomment ort job & remove specific branches

koganesi committed:
chore(ci): uncomment job dependency

koganesi committed:
chore(ci): Integration of FOSS check into the CI/CD pipeline (192) (#110)

koganesi committed:
chore: add apache 2.0 license to dev-project

koganesi committed:
chore: add apache 2.0 license to docs

koganesi committed:
feat(build): adapt build process to include shared assets in build-lifecycle (336) (#117)

koganesi committed:
chore(ci): move font insertion to FOSS Check pipeline

actions-user committed:
chore(ci): add SBOMs for workspaces

koganesi committed:
chore(ci): adjust insertion of google fonts (388) (#131)

koganesi committed:
chore: add apache 2.0 license to dev-project (387) (#118)

Jonas-Krahn committed:
Merge branch 'develop' into DEV-2/IAV-FF-chore-328

Jonas-Krahn committed:
Began fixing bugs

github-actions[bot] committed:
chore(format): apply automated formatting changes

Jonas-Krahn committed:
Rebase

github-actions[bot] committed:
chore(format): apply automated formatting changes

Jonas-Krahn committed:
Fixed Routing Bug

github-actions[bot] committed:
chore(format): apply automated formatting changes
